core/domain/html_validation_service.py

```diff
@@ -0,0 +1,890 @@
+        err_dict: dict. The dictionary which stores invalid tags and strings.
+
+    Returns:
+        bool. Boolean indicating whether a html string is valid for given RTE.
+    """
+    if rte_format == feconf.RTE_FORMAT_TEXTANGULAR:
+        RTE_TYPE = 'RTE_TYPE_TEXTANGULAR'
+    else:
+        RTE_TYPE = 'RTE_TYPE_CKEDITOR'
+    allowed_parent_list = feconf.RTE_CONTENT_SPEC[
+        RTE_TYPE]['ALLOWED_PARENT_LIST']
+    allowed_tag_list = feconf.RTE_CONTENT_SPEC[RTE_TYPE]['ALLOWED_TAG_LIST']
+
+    is_invalid = False
+
+    # Text with no parent tag is also invalid.
+    for content in soup.contents:
+        if not content.name:
+            is_invalid = True
+
+    for tag in soup.findAll():
+        # Checking for tags not allowed in RTE.
+        if tag.name not in allowed_tag_list:
+            if 'invalidTags' in err_dict:
+                err_dict['invalidTags'].append(tag.name)
+            else:
+                err_dict['invalidTags'] = [tag.name]
+            is_invalid = True
+        # Checking for parent-child relation that are not
+        # allowed in RTE.
+        parent = tag.parent.name
+        if (tag.name in allowed_tag_list) and (
+                parent not in allowed_parent_list[tag.name]):
+            if tag.name in err_dict:
+                err_dict[tag.name].append(parent)
+            else:
+                err_dict[tag.name] = [parent]
+            is_invalid = True
+
+    return is_invalid
+
+
+def add_caption_attr_to_image(html_string):
+    """Adds caption attribute to all oppia-noninteractive-image tags.
+
+    Args:
+        html_string. str: HTML string in which the caption attribute is to be
+            added.
+
+    Returns:
+        str. Updated HTML string with the caption attribute for all
+            oppia-noninteractive-image tags.
+    """
+    soup = bs4.BeautifulSoup(
+        html_string.encode(encoding='utf-8'), 'html.parser')
+
+    for image in soup.findAll(name='oppia-noninteractive-image'):
+        attrs = image.attrs
+        if 'caption-with-value' not in attrs:
+            image['caption-with-value'] = escape_html(json.dumps(''))
+
+    return unicode(soup)
+
+
+def validate_customization_args(html_list):
+    """Validates customization arguments of Rich Text Components in a list of
+    html string.
+
+    Args:
+        html_list: list(str). List of html strings to be validated.
+
+    Returns:
+        dict: Dictionary of all the invalid customisation args where
+            key is a Rich Text Component and value is the invalid html string.
+    """
+    # Dictionary to hold html strings in which customization arguments
+    # are invalid.
+    err_dict = {}
+    rich_text_component_list = [
+        'oppia-noninteractive-link',
+        'oppia-noninteractive-math',
+        'oppia-noninteractive-image',
+        'oppia-noninteractive-video',
+        'oppia-noninteractive-collapsible',
+        'oppia-noninteractive-tabs'
+    ]
+
+    for html_string in html_list:
```

Comment:
I think this would be a lot cleaner if this loop just finds the tags and puts them in a local list variable.

Then run a for loop over the list variable and pass each item to the second function in turn.

That way you have a "step 1" and a "step 2". If errors happen in either step it's easier to isolate the problem.

Reply:
Done!